Add script attributes from animation clips recursively recursively

aldelaro5 opened the pull request:
This allows to resolves attributes such as `middle.y` where middle is a Vector3 or anything serializable with a field named `y`.

I should point out: there doesn't seem to be anything preventing you from having a circular reference down the recursion, but I also couldn't find any game that was hanging from what I owned on steam. The best I got was it would add more to the cache, but they were valid paths.

If this is fine or there's a way to prevent this, let me know and I will do the necessary adjustments.

ds5678 commented:
Two things:

Infinite recursion is possible.

Use string interpolation instead of a StringBuilder. It's more efficient.

aldelaro5 commented:
> Four things:
> 
> Infinite recursion is possible.
> 
> Use string interpolation instead of a StringBuilder. It's more efficient.
> 
> There may be some additional optimizations using ReadOnlySpan.
> 
> Only primitive serialized types can be animated. There's a special C# type for this you can type check. Allocating strings for parent paths is unnecessary.

I partially addressed the review, sorry for taking so long.

The things left I need clarifications on are the parent string allocations and the ReadOnlySpan stuff. Specifically, what optimizations are you expecting with ReadOnlySpan and how do I avoid the string allocation if I need to append the . somewhere?

Additionally, while I did switched to string interpolation, I am curious why it's more efficient because I thought it would get lowered to Concat calls.

ds5678 reviewed (changes requested):
> while I did switched to string interpolation, I am curious why it's more efficient because I thought it would get lowered to Concat calls.

You are correct that it is the same as calling `string.Concat` in many situations, including this one. When I said that string interpolation was more efficient, that was specifically in comparison to a `StringBuilder`.

> parent string allocations

I addressed this in my suggested edit.

> what optimizations are you expecting with `ReadOnlySpan` and how do I avoid the string allocation if I need to append the . somewhere?

I am okay if you do nothing more in this regard. The optimizations significantly increase code verbosity. The general idea is that instead of allocating any strings, you use `ReadOnlySpan<char>` buffers to pass around the paths. The buffers can be created, either by stackalloc or renting from an `ArrayPool`.

> sorry for taking so long.

No worries. Thank you for coming back to it.

aldelaro5 commented on Source/AssetRipper.Processing/AnimationClips/PathChecksumCache.cs:
```diff
@@ -146,8 +146,27 @@ public void Add(IMonoScript script)
-			Add(field.Name);
+	private void AddFieldRecursively(SerializableType.Field field, string path, int depth = 0)
+	{
+		if (depth > 10)
```
It's arbitrary, you actually chose the number when i asked what timeout policy you preferred, but I'll specify it
